pkg/lockfile/lockfile_windows.go

```diff
@@ -22,48 +69,190 @@ import (
-	return &lockfile{locked: false}, nil
+	// Check if we can open the lock.
+	fd, err := openLock(path, ro)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error opening %q", path)
+	}
+	err = syscall.Close(fd)
+	if err != nil {
+		return nil, err
+	}
+
+	locktype := LOCKFILE_EXCLUSIVE_LOCK
+
+	if ro {
+		locktype = 0
+	}
+
+	return &lockfile{
+		stateMutex: &sync.Mutex{},
+		rwMutex:    &sync.RWMutex{},
+		file:       path,
+		lw:         stringid.GenerateRandomID(),
+		locktype:   locktype,
+		locked:     false,
+		ro:         ro}, nil
 }
 
 type lockfile struct {
-	mu     sync.Mutex
-	file   string
-	locked bool
+	// rwMutex serializes concurrent reader-writer acquisitions in the same process space
+	rwMutex *sync.RWMutex
+	// stateMutex is used to synchronize concurrent accesses to the state below
+	stateMutex *sync.Mutex
+	counter    int64
+	file       string
+	fd         syscall.Handle
+	lw         string
+	locktype   uint32
+	locked     bool
+	ro         bool
+	recursive  bool
 }
 
-func (l *lockfile) Lock() {
-	l.mu.Lock()
+func (l *lockfile) lock(lType uint32, recursive bool) {
+	switch lType {
+	case 0:
+		l.rwMutex.RLock()
+	case LOCKFILE_EXCLUSIVE_LOCK:
+		if recursive {
+			// NOTE: that's okay as recursive is only set in RecursiveLock(), so
+			// there's no need to protect against hypothetical RDLCK cases.
+			l.rwMutex.RLock()
+		} else {
+			l.rwMutex.Lock()
+		}
+	default:
+		panic(fmt.Sprintf("attempted to acquire a file lock of unrecognized type %d", lType))
+	}
+	l.stateMutex.Lock()
+	defer l.stateMutex.Unlock()
+	if l.counter == 0 {
+		// If we're the first reference on the lock, we need to open the file again.
+		fd, err := openLock(l.file, l.ro)
+		if err != nil {
+			panic(fmt.Sprintf("error opening %q: %v", l.file, err))
+		}
+		l.fd = fd
+
+		var ol syscall.Overlapped
+
+		// Optimization: only use the (expensive) fcntl syscall when
+		// the counter is 0.  In this case, we're either the first
+		// reader lock or a writer lock.
+		ret, _, err := procLockFileEx.Call(uintptr(l.fd), uintptr(lType), 0, uintptr(MAXDWORD), uintptr(MAXDWORD), uintptr(unsafe.Pointer(&ol)))
+		if ret == 0 {
+			return
+		}
+	}
+	l.locktype = lType
 	l.locked = true
+	l.recursive = recursive
+	l.counter++
+}
+
+// Lock locks the lockfile as a writer.  Panic if the lock is a read-only one.
+func (l *lockfile) Lock() {
+	if l.ro {
+		panic("can't take write lock on read-only lock file")
+	} else {
+		l.lock(LOCKFILE_EXCLUSIVE_LOCK, false)
+	}
 }
 
 func (l *lockfile) RecursiveLock() {
-	// We don't support Windows but a recursive writer-lock in one process-space
-	// is really a writer lock, so just panic.
-	panic("not supported")
+	if l.ro {
+		l.RLock()
+	} else {
+		l.lock(LOCKFILE_EXCLUSIVE_LOCK, true)
+	}
 }
 
 func (l *lockfile) RLock() {
-	l.mu.Lock()
-	l.locked = true
+	l.lock(0, false)
 }
 
 func (l *lockfile) Unlock() {
-	l.locked = false
-	l.mu.Unlock()
+	l.stateMutex.Lock()
```

Comment:
The cut and paste code from lockfile_unix.go should be moved to lockfile.go and removed from lockfile_windows.go and lockfile_unix.go.